File: Baekjoon/Class3/46.7662 이중 우선순위 큐.cpp
// priority_queue 사용
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

void run_dual_priority_queue() {
    int k;
    cin >> k;

    // <num, index>
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> min_heap;
    priority_queue<pair<int, int>> max_heap;
    vector<char> visited(k);

    for (int i = 0; i < k; i++) {
        char operation;
        int num;
        cin >> operation >> num;

        switch (operation) {
        case 'I': // 삽입
            min_heap.push({ num, i });
            max_heap.push({ num, i });
            break;
        case 'D':
            if (num == -1) { // 최솟값 삭제
                while (!min_heap.empty()) {
                    if (visited[min_heap.top().second]) { // 방문했던 노드 삭제
                        min_heap.pop();
                        continue;
                    }
                    visited[min_heap.top().second] = true;
                    min_heap.pop();
                    break;
                }
            }
            else { // 최댓값 삭제
                while (!max_heap.empty()) {
                    if (visited[max_heap.top().second]) { // 방문했던 노드 삭제
                        max_heap.pop();
                        continue;
                    }
                    visited[max_heap.top().second] = true;
                    max_heap.pop();
                    break;
                }
            }
        }
    }
    while (!min_heap.empty() && visited[min_heap.top().second]) min_heap.pop();
    while (!max_heap.empty() && visited[max_heap.top().second]) max_heap.pop();

    if (min_heap.empty())
        cout << "EMPTY\n";
    else
        cout << max_heap.top().first << ' ' << min_heap.top().first << '\n'; // 최댓값 최솟값 출력
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;
    while (t--)
        run_dual_priority_queue();
    return 0;
}

/* multiset 사용
#include <iostream>
#include <set>
using namespace std;

void run_dual_priority_queue() {
    int k;
    cin >> k;

    multiset<int> q; // 이중 우선순위 큐

    while (k--) {
        char operation;
        int num;
        cin >> operation >> num;

        switch (operation) {
        case 'I': // 삽입
            q.insert(num);
            break;
        case 'D':
            if (q.empty())
                break;
            if (num == -1) // 최솟값 삭제
                q.erase(q.begin());
            else // 최댓값 삭제
                q.erase(--q.end());
        }
    }

    if (q.empty())
        cout << "EMPTY\n";
    else
        cout << *q.rbegin() << ' ' << *q.begin() << '\n'; // 최댓값 최솟값 출력
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;
    while (t--)
        run_dual_priority_queue();
    return 0;
}
*/

File: Baekjoon/Class3/46.7662 이중 우선순위 큐.java
// PriorityQueue 사용
import java.io.*;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main {
    private static class Pair {
        int num, index;
        public Pair(int num, int index) { this.num = num; this.index = index; }
    }

    private static void runDualPriorityQueue(BufferedReader br, StringBuilder sb) throws IOException {
        int k = Integer.parseInt(br.readLine());

        PriorityQueue<Pair> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a.num, b.num));
        PriorityQueue<Pair> maxHeap = new PriorityQueue<>((a, b) -> Integer.compare(b.num, a.num));
        boolean[] visited = new boolean[k];

        for(int i = 0; i < k; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String operation = st.nextToken();
            int num = Integer.parseInt(st.nextToken());

            switch(operation.charAt(0)) {
                case 'I': // 삽입
                    minHeap.add(new Pair(num, i));
                    maxHeap.add(new Pair(num, i));
                    break;
                case 'D':
                    if(num == -1) { // 최솟값 삭제
                        while(!minHeap.isEmpty()) {
                            if(visited[minHeap.peek().index]) { // 방문했던 노드 삭제
                                minHeap.remove();
                                continue;
                            }
                            visited[minHeap.peek().index] = true;
                            minHeap.remove();
                            break;
                        }
                    }
                    else { // 최댓값 삭제
                        while(!maxHeap.isEmpty()) {
                            if(visited[maxHeap.peek().index]) { // 방문했던 노드 삭제
                                maxHeap.remove();
                                continue;
                            }
                            visited[maxHeap.peek().index] = true;
                            maxHeap.remove();
                            break;
                        }
                    }
            }
        }
        while(!minHeap.isEmpty() && visited[minHeap.peek().index]) minHeap.remove();
        while(!maxHeap.isEmpty() && visited[maxHeap.peek().index]) maxHeap.remove();

        if(minHeap.isEmpty())
            sb.append("EMPTY\n");
        else
            sb.append(maxHeap.peek().num).append(' ').append(minHeap.peek().num).append('\n'); // 최댓값 최솟값 출력
    }

    public static void main(String[] args) throws IOException {
        int t;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        t = Integer.parseInt(br.readLine());
        for(int i = 0; i < t; i++)
            runDualPriorityQueue(br, sb);
        System.out.print(sb);
        br.close();
    }
}

/* TreeMap 사용
import java.io.*;
import java.util.StringTokenizer;
import java.util.TreeMap;

class Main {
    private static void runDualPriorityQueue(BufferedReader br, StringBuilder sb) throws IOException {
        int k = Integer.parseInt(br.readLine());

        TreeMap<Integer, Integer> q = new TreeMap<>(); // <num, count>
        for(int i = 0; i < k; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String operation = st.nextToken();
            int num = Integer.parseInt(st.nextToken());

            switch(operation.charAt(0)) {
                case 'I': // 삽입
                    q.put(num, q.getOrDefault(num, 0) + 1);
                    break;
                case 'D': // 삭제
                    if(q.isEmpty())
                        break;
                    int numToRemove = (num == -1) ? q.firstKey() : q.lastKey(); // 최댓값 : 최솟값
                    int count = q.get(numToRemove);
                    if(count == 1)
                        q.remove(numToRemove);
                    else
                        q.put(numToRemove, count - 1);
            }
        }

        if(q.isEmpty())
            sb.append("EMPTY\n");
        else
            sb.append(q.lastKey()).append(' ').append(q.firstKey()).append('\n');
    }

    public static void main(String[] args) throws IOException {
        int t;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        t = Integer.parseInt(br.readLine());
        for(int i = 0; i < t; i++)
            runDualPriorityQueue(br, sb);
        System.out.print(sb);
        br.close();
    }
}
*/